server/src/main/java/com/cloud/user/AccountManagerImpl.java

```diff
@@ -1110,6 +1183,39 @@ public UserAccount createUserAccount(final String userName, final String passwor
         return _userAccountDao.findById(userId);
     }
 
+    /**
+     * if there is any permission under the requested role that is not permitted for the caller, refuse
+     */
+    private void checkRoleEscalation(Account caller, Account requested) {
+        Long requestedRoleId = requested.getRoleId();
+        List<RolePermission> requestedPermissions = roleService.findAllPermissionsBy(requestedRoleId);
+        for (String command : apiNameList) {
```

Comment:
Could there be a simpler fix, that is the caller is only allowed to select/pass roleids while creating account that is equal or lower than their own role; for example;
root admin -> can do anything
domain admin -> create domain/sub-domain admins and user-accounts
user accounts -> none

Reply:
that would not be taking into account dynamic roles. We want the caller not to be able to create an account with a role that has any permission that they don't have themselves, independent of the RoleType.